src/databricks/labs/ucx/assessment/sequencing.py

```diff
@@ -0,0 +1,352 @@
+        """Convert to class:MigrationStep."""
+        return MigrationStep(
+            step_id=self.node_id,
+            step_number=step_number,
+            object_type=self.object_type,
+            object_id=self.object_id,
+            object_name=self.object_name,
+            object_owner=self.object_owner,
+            required_step_ids=required_step_ids,
+        )
+
+
+@dataclass
+class MaybeMigrationNode:
+    node: MigrationNode | None
+    problems: list[DependencyProblem]
+
+    @property
+    def failed(self) -> bool:
+        return len(self.problems) > 0
+
+
+# We expect `tuple[int, int, MigrationNode | str]`
+# for `[priority, counter, MigrationNode | PriorityQueue._REMOVED | PriorityQueue_UPDATED]`
+# but we use list for the required mutability
+QueueEntry = list[int | MigrationNode | str]
+
+
+class PriorityQueue:
+    """A migration node priority queue.
+
+    Note:
+        This implementation does not support threading safety as that is not required.
+
+    Source:
+        See https://docs.python.org/3/library/heapq.html#priority-queue-implementation-notes on the changes below
+        to handle priority changes in the task. Also, the _UPDATED marker is introduced to avoid updating removed nodes.
+    """
+
+    _REMOVED = "<removed>"  # Mark removed items
+
+    def __init__(self):
+        self._entries: list[QueueEntry] = []
+        self._entry_finder: dict[MigrationNode, QueueEntry] = {}
+        self._counter = itertools.count()  # Tiebreaker with equal priorities, then "first in, first out"
+
+    def put(self, priority: int, task: MigrationNode) -> None:
+        """Put or update task in the queue.
+
+        The lowest priority is retrieved from the queue first.
+        """
+        if task in self._entry_finder:
+            self._remove(task)
+        count = next(self._counter)
+        entry = [priority, count, task]
+        self._entry_finder[task] = entry
+        heapq.heappush(self._entries, entry)
+
+    def get(self) -> MigrationNode | None:
+        """Gets the tasks with the lowest priority."""
+        while self._entries:
+            _, _, task = heapq.heappop(self._entries)
+            if task == self._REMOVED:
+                continue
+            assert isinstance(task, MigrationNode)
+            self._remove(task)
+            return task
+        return None
+
+    def _remove(self, task: MigrationNode) -> None:
+        """Remove a task from the queue."""
+        entry = self._entry_finder.pop(task)
+        entry[2] = self._REMOVED
+
+
+class MigrationSequencer:
+    """Sequence the migration dependencies in order to execute the migration.
+
+    Similar to the other graph logic, we first build the graph by registering dependencies, then we analyse the graph.
+    Analysing the graph in this case means: computing the migration sequence in `meth:generate_steps`.
+    """
+
+    def __init__(self, ws: WorkspaceClient, admin_locator: AdministratorLocator):
```

Comment:
```suggestion
    def __init__(self, ws: WorkspaceClient, admininstrator_locator: AdministratorLocator):
```
properties have to be named consistently - makes it easier to refactor.